MAINT: Translate binsearch.c.src to C++ using templates.

serge-sans-paille opened the pull request:
Side effects:
- make each type tag aware of how to compare them
- add a few __cplusplus guard in numpy header
- handle confict between some static and non static symbols with the same name

serge-sans-paille commented:
The Pypy error seems unrelated :-/

mattip commented:
> The Pypy error seems unrelated :-/

xref #20491, probably picking up an unstable nightly build

scratchmex commented:
Could `template` of C++ replace most of the `.c.src` code?

serge-sans-paille commented:
@scratchmex that's indeed the goal, see #19713

seiko2plus commented on numpy/core/src/npysort/binsearch.cpp:
```diff
@@ -81,31 +63,32 @@ binsearch_@side@_@suff@(const char *arr, const char *key, char *ret,
                            npy_intp arr_str, npy_intp key_str,
                            npy_intp sort_str, npy_intp ret_str,
-                           PyArrayObject *NOT_USED)
+                           Cmp cmp)
```
```suggestion
```
static pointers require at least optimization level 2 in order to be inlined, it's better to use lambda or just pass meta-type as template arg holds `operator ()` contains the scalar operation. for example:
```c++


template<typename Tag>
struct cmp_lt {
    template <typename T>
    bool operator () (T a, T b)
    { return a < b; }
};

template<>
struct cmp_lt<half_tag> {
    bool operator () (npy_half a, npy_half b)
    { return HALF_LT(a, b); }
};
// specialize the rest float/cfloat
// ...
```
to pass it
```C++
argbinsearch_<half_tag, cmp_lt>(...);
```
to initialize it
```C++
template<class Tag, typename Cmp, class T = typename Tag::type>
static int argbinsearch_(...)
{
    Cmp<Tag> cmp;
}
```

serge-sans-paille replied:
This is actually very compiler specific, see https://godbolt.org/z/7TGKvM8o3
I've yet to see a case where -O1 is used in a place where optimization really matters, so I'm not a big fan of this argument. at -O2 all compiler inline the function pointer correctly, and I tend to prefer readability over optimizing for -O1. What are the thoughts of the other reviewers on that point?

charris replied:
I can see `-O1` being used for debugging. `-O3` breaks occsionally, but `-O2` usually fixes it. That said, having unstated assumptions floating around could be dangerous. Is it the case that inlining is not required for the code to run correctly? If so, might add a comment

serge-sans-paille replied:
It doesn't change the smeantic of the code, it only changes inlining, i.e. execution speed.

seiko2plus replied:
> This is actually very compiler specific, see https://godbolt.org/z/7TGKvM8o3

on GCC `-O1` https://godbolt.org/z/4b6f3hxcM isn't going to inline static pointers.

> I tend to prefer readability over optimizing

passing a function pointer for internal use that can be handled in compile time is 100% against readability, 100% against the idea behind the existence of C++. The above suggestion is a common pattern we tend to use it since C++ 98 and it can be replaced with lambda.

@charris,
> I can see -O1 being used for debugging

we use `-O0` for debugging, distros may build NumPy with `-O1`.

> It doesn't change the smeantic of the code, it only changes inlining, i.e. execution speed.

giving a chance for the compiler to avoid inlining a function with one line expression that can be executed within a loop is in general a sign of bad design.

serge-sans-paille replied:
I thought `BOOL_LT` was used else where, which is why I reused existing symbol: to prevent duplication. it turns out it's not the case, so I'll move the definition to the tags.

serge-sans-paille replied:
It's actually used in heap sort. To avoid duplication, I'll keep the current implementation until it's not needed elsewhere. Small steps!

seiko2plus replied:
adding `extern C` to [npysort_common.h](https://github.com/numpy/numpy/blob/main/numpy/core/src/npysort/npysort_common.h) will fix the duplication between C and C++ symbols without the need of wraps around via static pointers.

serge-sans-paille commented on numpy/core/src/npysort/binsearch.cpp:
```diff
@@ -247,4 +223,57 @@ npy_argbinsearch_@side@(const char *arr, const char *key,
+  DEFINE_NUMERIC_SEARCHES(left, int, less)
+  DEFINE_NUMERIC_SEARCHES(left, uint, less)
+  DEFINE_NUMERIC_SEARCHES(left, long, less)
+  DEFINE_NUMERIC_SEARCHES(left, ulong, less)
+  DEFINE_NUMERIC_SEARCHES(left, longlong, less)
+  DEFINE_NUMERIC_SEARCHES(left, ulonglong, less)
+  DEFINE_NUMERIC_SEARCHES(left, half, less)
+  DEFINE_NUMERIC_SEARCHES(left, float, less)
+  DEFINE_NUMERIC_SEARCHES(left, double, less)
+  DEFINE_NUMERIC_SEARCHES(left, longdouble, less)
+  DEFINE_NUMERIC_SEARCHES(left, cfloat, less)
+  DEFINE_NUMERIC_SEARCHES(left, cdouble, less)
+  DEFINE_NUMERIC_SEARCHES(left, clongdouble, less)
+  DEFINE_NUMERIC_SEARCHES(left, datetime, less)
+  DEFINE_NUMERIC_SEARCHES(left, timedelta, less)
+
+  DEFINE_NUMERIC_SEARCHES(right, bool, less_equal)
+  DEFINE_NUMERIC_SEARCHES(right, byte, less_equal)
+  DEFINE_NUMERIC_SEARCHES(right, ubyte, less_equal)
+  DEFINE_NUMERIC_SEARCHES(right, short, less_equal)
+  DEFINE_NUMERIC_SEARCHES(right, ushort, less_equal)
+  DEFINE_NUMERIC_SEARCHES(right, int, less_equal)
+  DEFINE_NUMERIC_SEARCHES(right, uint, less_equal)
+  DEFINE_NUMERIC_SEARCHES(right, long, less_equal)
+  DEFINE_NUMERIC_SEARCHES(right, ulong, less_equal)
+  DEFINE_NUMERIC_SEARCHES(right, longlong, less_equal)
+  DEFINE_NUMERIC_SEARCHES(right, ulonglong, less_equal)
+  DEFINE_NUMERIC_SEARCHES(right, half, less_equal)
+  DEFINE_NUMERIC_SEARCHES(right, float, less_equal)
+  DEFINE_NUMERIC_SEARCHES(right, double, less_equal)
+  DEFINE_NUMERIC_SEARCHES(right, longdouble, less_equal)
+  DEFINE_NUMERIC_SEARCHES(right, cfloat, less_equal)
+  DEFINE_NUMERIC_SEARCHES(right, cdouble, less_equal)
+  DEFINE_NUMERIC_SEARCHES(right, clongdouble, less_equal)
+  DEFINE_NUMERIC_SEARCHES(right, datetime, less_equal)
+  DEFINE_NUMERIC_SEARCHES(right, timedelta, less_equal)
+
+  DEFINE_GENERIC_SEARCHES(left, std::less<int>)
+  DEFINE_GENERIC_SEARCHES(right, std::less_equal<int>)
+}
```
I 100% agree with the approach. i'll give it a try, and depending on how invasive it is in the code base, I'll add that to this commit or submit another PR.

seiko2plus replied:
either this approach or you can bring the old template repeater to handle this situation instead of using macros.

serge-sans-paille replied:
@seiko2plus : new version that implements your suggested approach. I'm *really* happy with the result. It makes the C to C++ step a bit higher, but the result really looks great (according to me ;-))
Thanks for suggesting that path o/

serge-sans-paille commented:
Folks, any thoughts on this new version?

serge-sans-paille committed:
Convert binsearch.c.src to binsearch.cpp

Side effects:
- make each type tag aware of how to compare them
- add a few __cplusplus guard in numpy header
- handle confict between some static and non static symbols with the same name

serge-sans-paille commented:
(rebased on main)

seiko2plus commented on numpy/core/src/npysort/binsearch.cpp:
```diff
@@ -0,0 +1,371 @@
+ *****************************************************************************
+ */
+
+template<arg_t arg>
+struct binsearch_base;
+
+template<>
+struct binsearch_base<arg> {
+  using function_type = PyArray_ArgBinSearchFunc*;
+  struct value_type {
+    int typenum;
+    function_type binsearch[NPY_NSEARCHSIDES];
+  };
+  template<class... Tags>
+  static constexpr std::array<value_type, sizeof...(Tags)> make_binsearch_map(npy::taglist<Tags...>) {
+    return std::array<value_type, sizeof...(Tags)>{value_type{Tags::type_value, { (function_type)&argbinsearch<Tags, left>,(function_type)argbinsearch<Tags, right>}}...};
+  }
```
Nice !, would you please sort out the code style here?

seiko2plus commented:
@serge-sans-paille,
> Folks, any thoughts on this new version?

It looks way better than the previous one, easier to read, no more C macros. Good job!.

serge-sans-paille commented:
@seiko2plus updated with formatting and usage of wrapper instead of function pointers

mattip commented:
I wonder why coverage thinks some of the code paths are not covered. Is that a false-positive?

serge-sans-paille commented:
@mattip I think it indicates that binsearch is not tested for all scalar types - the comparison functions are used elsewhere in the C code base so their coverage is 100%, but the C++ version are (currently) only used in binsearch.

Two paths here: either we test all dtypes for binsearch, or we wait until I convert other sort algorithms. Your call ;-)

mattip commented:
I think we can wait for the rest of sort. @seiko2plus any reason not to merge?

charris commented:
Do we want to run clang-format on the files? We have the needed `.clang-format`.

charris commented:
For test coverage, there are likely many numpy tests that could be improved by testing over more dtypes.

seiko2plus commented:
@mattip, I have some minor reservations, not sure if it deserves another round of review. In general, the new code looks good to me and we should merge it.

mattip commented:
Thanks @serge-sans-paille
